controller/node_controller.go

```diff
@@ -356,19 +370,67 @@ func (nc *NodeController) syncNode(key string) (err error) {
-			nc.eventRecorder.Eventf(node, v1.EventTypeWarning, types.NodeConditionReasonKubernetesNodeDown, "Kubernetes node missing: node %v has been removed from the cluster and there is no manager pod running on it", node.Name)
+			nc.eventRecorder.Eventf(node, v1.EventTypeWarning, types.NodeConditionReasonManagerPodMissing, "manager pod missing: node %v has no manager pod running on it", node.Name)
 		}
 		condition.Status = types.ConditionStatusFalse
-		condition.Reason = string(types.NodeConditionReasonKubernetesNodeDown)
-		condition.Message = fmt.Sprintf("Kubernetes node missing: node %v has been removed from the cluster and there is no manager pod running on it", node.Name)
+		condition.Reason = string(types.NodeConditionReasonManagerPodMissing)
+		condition.Message = fmt.Sprintf("manager pod missing: node %v has no manager pod running on it", node.Name)
 		node.Status.Conditions[types.NodeConditionTypeReady] = condition
-		// set node unschedulable
-		node.Spec.AllowScheduling = false
+	}
+
+	// sync node state with kuberentes node status
+	kubeNode, err := nc.ds.GetKubernetesNode(name)
+	if err != nil {
+		// if kubernetes node has been removed from cluster
+		if apierrors.IsNotFound(err) {
+			condition := types.GetNodeConditionFromStatus(node.Status, types.NodeConditionTypeReady)
+			if condition.Status != types.ConditionStatusFalse {
+				condition.LastTransitionTime = util.Now()
+				nc.eventRecorder.Eventf(node, v1.EventTypeWarning, types.NodeConditionReasonKubernetesNodeDown, "Kubernetes node missing: node %v has been removed from the cluster and there is no manager pod running on it", node.Name)
+			}
+			condition.Status = types.ConditionStatusFalse
+			condition.Reason = string(types.NodeConditionReasonKubernetesNodeDown)
+			condition.Message = fmt.Sprintf("Kubernetes node missing: node %v has been removed from the cluster and there is no manager pod running on it", node.Name)
+			node.Status.Conditions[types.NodeConditionTypeReady] = condition
+			// set node unschedulable
+			node.Spec.AllowScheduling = false
+		} else {
+			return err
+		}
+	} else {
+		kubeConditions := kubeNode.Status.Conditions
+		for _, con := range kubeConditions {
+			if con.Type == v1.NodeReady {
+				if con.Status != v1.ConditionTrue {
+					condition := types.GetNodeConditionFromStatus(node.Status, types.NodeConditionTypeReady)
+					if condition.Status != types.ConditionStatusFalse {
+						condition.LastTransitionTime = util.Now()
+						nc.eventRecorder.Eventf(node, v1.EventTypeWarning, types.NodeConditionReasonKubernetesNodeNotReady, "Kubernetes node %v not ready: %v", node.Name, con.Reason)
+					}
+					condition.Status = types.ConditionStatusFalse
+					condition.Reason = string(types.NodeConditionReasonKubernetesNodeNotReady)
+					condition.Message = fmt.Sprintf("Kubernetes node %v not ready: %v", node.Name, con.Reason)
+					node.Status.Conditions[types.NodeConditionTypeReady] = condition
+				}
+			} else {
```

Comment:
We cannot treat all others blindly as the all negative conditions. Use switch here for specific conditions. We may treat them the same way, but we cannot assume that all the conditions are negative except NodeReady. There are can be more positive conditions coming in the future releases.